vectorstores/alloydb/engine.go

```diff
@@ -0,0 +1,148 @@
+package alloydb
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+
+	"cloud.google.com/go/alloydbconn"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/idtoken"
+	"google.golang.org/api/oauth2/v2"
+	"google.golang.org/api/option"
+)
+
+type EmailRetreiver func(context.Context) (string, error)
+
+type PostgresEngine struct {
+	conn *pgxpool.Pool
+}
+
+// NewPostgresEngine creates a new PostgresEngine.
+func NewPostgresEngine(ctx context.Context, opts ...Option) (*PostgresEngine, error) {
+	pgEngine := new(PostgresEngine)
+	cfg := &engineConfig{
+		port:           defaultPort,
+		emailRetreiver: getServiceAccountEmail,
+	}
+	for _, opt := range opts {
+		opt(cfg)
+	}
+	username, usingIAMAuth, err := getUser(ctx, *cfg)
+	if err != nil {
+		return nil, fmt.Errorf("error assigning user. Err: %w", err)
+	}
+	if usingIAMAuth {
+		token, err := getIAMToken(ctx, username)
+		if err != nil {
+			return nil, err
+		}
+		cfg.password = token
+	}
+	connPool, err := createConnection(ctx, *cfg)
+	if err != nil {
+		return &PostgresEngine{}, err
+	}
+	pgEngine.conn = connPool
+	return pgEngine, nil
+}
+
+// createConnection creates a connection pool to the PostgreSQL database.
+func createConnection(ctx context.Context, cfg engineConfig) (*pgxpool.Pool, error) {
+	// Configure the driver to connect to the database
+	dsn := fmt.Sprintf("user=%s password=%s dbname=%s sslmode=disable", cfg.user, cfg.password, cfg.database)
+	config, err := pgxpool.ParseConfig(dsn)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse connection config: %w", err)
+	}
+
+	// Create a new dialer with any options
+	d, err := alloydbconn.NewDialer(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize connection: %w", err)
+	}
+
+	instanceAddress := fmt.Sprintf("projects/%s/locations/%s/clusters/%s/instances/%s", cfg.projectID, cfg.region, cfg.cluster, cfg.instance)
+	// Create the connection
+	config.ConnConfig.DialFunc = func(ctx context.Context, _ string, instanceAddr string) (net.Conn, error) {
+		return d.Dial(ctx, instanceAddress)
+	}
+	conn, err := pgxpool.NewWithConfig(ctx, config)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create connection pool: %w", err)
+	}
+	return conn, nil
+}
+
+// Close closes the connection.
+func (p *PostgresEngine) Close() error {
```

Comment:
nit: not super sure this should return an error -- if conn is nil I would consider already closed :)